### PR TITLE
Feature - add caching functionality to secret store

### PR DESCRIPTION
**Pull request by stijnmoreels**
Relates to #137

### CODE REVIEW

**Comment by stijnmoreels** [comment]
Waiting for #134 for docs on this.

**Comment by stijnmoreels** [comment]
Updated cached provider with seperate caching configuration for each provider.

**Comment by tomkerkhove** [comment]
I think the reasoning is ok but would scope it per provider where we only need it for Azure Key Vault.

This could be something similar to this:
```csharp
public class Program
{
    public static void Main(string[] args)
    {
        CreateHostBuilder(args).Build().Run();
    }

    public static IHostBuilder CreateHostBuilder(string[] args) =>
        Host.CreateDefaultBuilder(args)
            .ConfigureHostConfiguration(configBuilder =>
            {
                configBuilder.AddJsonFile("appsettings.Development.json")
                             .AddJsonFile("appsettings.json");
            })
            .ConfigureSecretStore((context, config, builder) =>
            {
#if DEBUG
                builder.AddConfiguration(config);
#endif
                var keyVaultName = config["KeyVault_Name"];
                builder.AddEnvironmentVariables()
                       .AddAzureKeyVaultWithManagedServiceIdentity($"https://{keyVaultName}.vault.azure.net", allowCaching: true);
                       .AddAzureKeyVaultWithManagedServiceIdentity($"https://{keyVaultName}.vault.azure.net", cacheConfiguration);
            })
            .ConfigureWebHostDefaults(webBuilder =>
            {
                webBuilder.UseStartup<Startup>();
            });
}
```

**Comment by stijnmoreels** [comment]
> I think the reasoning is ok but would scope it per provider where we only need it for Azure Key Vault.
> 
> This could be something similar to this:
> 
> ```cs
> public class Program
> {
>     public static void Main(string[] args)
>     {
>         CreateHostBuilder(args).Build().Run();
>     }
> 
>     public static IHostBuilder CreateHostBuilder(string[] args) =>
>         Host.CreateDefaultBuilder(args)
>             .ConfigureHostConfiguration(configBuilder =>
>             {
>                 configBuilder.AddJsonFile("appsettings.Development.json")
>                              .AddJsonFile("appsettings.json");
>             })
>             .ConfigureSecretStore((context, config, builder) =>
>             {
> #if DEBUG
>                 builder.AddConfiguration(config);
> #endif
>                 var keyVaultName = config["KeyVault_Name"];
>                 builder.AddEnvironmentVariables()
>                        .AddAzureKeyVaultWithManagedServiceIdentity($"https://{keyVaultName}.vault.azure.net", allowCaching: true);
>                        .AddAzureKeyVaultWithManagedServiceIdentity($"https://{keyVaultName}.vault.azure.net", cacheConfiguration);
>             })
>             .ConfigureWebHostDefaults(webBuilder =>
>             {
>                 webBuilder.UseStartup<Startup>();
>             });
> }
> ```

Yes, updated that.

File: src/Arcus.Security.Core/SecretStoreBuilder.cs [hunk 1]
```diff
@@ -36,9 +52,23 @@ public SecretStoreBuilder(IServiceCollection services)
         public SecretStoreBuilder AddProvider(ISecretProvider secretProvider)
         {
             Guard.NotNull(secretProvider, nameof(secretProvider));
-            Services.AddSingleton(new SecretStoreSource(secretProvider));
+            SecretStoreSources.Add(new SecretStoreSource(secretProvider));
```
[the remark this answers is not in the document]
Comment (stijnmoreels) [reply src/Arcus.Security.Core/SecretStoreBuilder.cs:55]: No, the root secret provider is registered as both cached and non cached, and is directing calls to the child registrations, non-cached calls will not result in cached calls.

File: src/Arcus.Security.Providers.AzureKeyVault/Configuration/SecretStoreBuilderExtensions.cs [hunk 2]
```diff
@@ -63,11 +77,15 @@ public static class SecretStoreBuilderExtensions
         /// <param name="rawVaultUri">The Uri of the Azure Key Vault you want to connect to.</param>
         /// <param name="clientId">The ClientId of the service principal, used to connect to Azure Key Vault</param>
         /// <param name="clientKey">The Secret ClientKey of the service principal, used to connect to Azure Key Vault</param>
+        /// <param name="allowCaching">The flag to indicate whether to include caching during secret retrieval in Azure key vault.</param>
+        /// <param name="cacheConfiguration">The configuration to control how the caching will be done (setting this will automatically enable <paramref name="allowCaching"/>).</param>
         public static SecretStoreBuilder AddAzureKeyVaultWithServicePrincipal(
```
[the remark this answers is not in the document]
Comment (stijnmoreels) [reply src/Arcus.Security.Providers.AzureKeyVault/Configuration/SecretStoreBuilderExtensions.cs:82]: Hmm, 
I don't find it very useful then to have an overload where you specify you want caching but don't know when the caching will happen.

[the remark this answers is not in the document]
Comment (tomkerkhove) [reply src/Arcus.Security.Providers.AzureKeyVault/Configuration/SecretStoreBuilderExtensions.cs:82]: It's control over simplicity - If you don't configure how the cache behaves you are fine with the default mechanics and don't care about it. THose who want, however, can tweak it but for most of the cases you don't care.

[the remark this answers is not in the document]
Comment (stijnmoreels) [reply src/Arcus.Security.Providers.AzureKeyVault/Configuration/SecretStoreBuilderExtensions.cs:82]: Updated with proposal.

File: src/Arcus.Security.Providers.AzureKeyVault/Configuration/SecretStoreBuilderExtensions.cs [hunk 3]
```diff
@@ -42,18 +73,113 @@ public static class SecretStoreBuilderExtensions
...
+                allowCaching);
+        }
+
+        /// <summary>
+        /// Adds Azure Key Vault as a secret source which uses Managed Identity authentication.
+        /// </summary>
+        /// <param name="builder">The builder to create the secret store.</param>
+        /// <param name="rawVaultUri">The Uri of the Azure Key Vault you want to connect to.</param>
+        /// <param name="cacheConfiguration">The configuration to control how the caching will be done.</param>
+        public static SecretStoreBuilder AddAzureKeyVaultWithManagedServiceIdentity(
```
Comment (tomkerkhove) [inline src/Arcus.Security.Providers.AzureKeyVault/Configuration/SecretStoreBuilderExtensions.cs:99]: I'm thinking out loud but shouldn't we align with the overload above and use optional except for the cache configuration?

Ok with splitting into dedicated methods, but then we should maybe do something similar for the one above?

Comment (stijnmoreels) [reply src/Arcus.Security.Providers.AzureKeyVault/Configuration/SecretStoreBuilderExtensions.cs:99]: Move back to one and moved the cache config before the optional params.

**Pull request merged**